### PR TITLE
K8SPSMDB-305: Make possible to configure storage for each Replica Set

### PR DESCRIPTION
**Pull request by blez**
[K8SPSMDB-305](https://jira.percona.com/browse/K8SPSMDB-305)

### CODE REVIEW

**Review by tplavcic — DISMISSED** [review]

File: pkg/apis/psmdb/v1/psmdb_defaults.go [hunk 1]
```diff
@@ -157,6 +157,10 @@ func (cr *PerconaServerMongoDB) CheckNSetDefaults(platform version.Platform, log
 			return errors.New("mongos should be specified")
 		}
```
Comment (delgod) [inline pkg/apis/psmdb/v1/psmdb_defaults.go:159]: can we also add check about configSrv engine (as was asked in ticket)
```
if cr.Spec.Sharding.ConfigsvrReplSet.Storage.Engine != WiredTiger {
    return errors.Errorf("%s storage engine is not supported for config server", cr.Spec.Sharding.ConfigsvrReplSet.Storage.Engine)
}

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/apis/psmdb/v1/psmdb_defaults.go (hunk 1 above)
Comment (blez) [reply pkg/apis/psmdb/v1/psmdb_defaults.go:159]: Engine is set only here, in mongod section

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-467-d680454c

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| init-deploy | passed |
| monitoring | passed |
| self-healing | passed |
| upgrade-consistency | passed |
| limits | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| scaling | passed |
| operator-self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

File: pkg/apis/psmdb/v1/psmdb_defaults.go (hunk 1 above)
Comment (delgod) [reply pkg/apis/psmdb/v1/psmdb_defaults.go:159]: in such case, can we always force WiredTiger for configSrv replicaset?
don't take value from mongod section?

Comment (blez) [reply pkg/apis/psmdb/v1/psmdb_defaults.go:159]: Config Replica set must to use WiredTiger
https://docs.mongodb.com/manual/core/sharded-cluster-config-servers/#replica-set-config-servers
```
To deploy config servers as a replica set, the config servers must run the WiredTiger storage engine.
```

File: deploy/cr.yaml [hunk 2]
```diff
@@ -120,6 +137,20 @@ spec:
 
     configsvrReplSet:
       size: 3
+#      storage:
+#        engine: wiredTiger
+#        inMemory:
+#          engineConfig:
+#            inMemorySizeRatio: 0.9
```
Comment (delgod) [inline deploy/cr.yaml:144]: can we remove these 3 lines?

File: pkg/apis/psmdb/v1/psmdb_defaults.go [hunk 3]
```diff
@@ -232,6 +202,45 @@ func (cr *PerconaServerMongoDB) CheckNSetDefaults(platform version.Platform, log
 	}
```
Comment (delgod) [inline pkg/apis/psmdb/v1/psmdb_defaults.go:203]: should we also add?
```
if cr.Spec.Sharding.ConfigsvrReplSet.Storage.Engine != WiredTiger {
    return errors.Errorf("%s storage engine is not supported for config server", cr.Spec.Sharding.ConfigsvrReplSet.Storage.Engine)
}
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| demand-backup-sharded | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-467-08a70aa8

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

File: pkg/apis/psmdb/v1/psmdb_defaults.go (hunk 3 above)
Comment (blez) [reply pkg/apis/psmdb/v1/psmdb_defaults.go:203]: https://github.com/percona/percona-server-mongodb-operator/pull/467/commits/a3d77d79a78b9eedca1a3d9b409f9101c4a427fb

File: deploy/cr.yaml (hunk 2 above)
Comment (blez) [reply deploy/cr.yaml]: https://github.com/percona/percona-server-mongodb-operator/pull/467/commits/a3d77d79a78b9eedca1a3d9b409f9101c4a427fb

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-467-a3d77d79

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Review by tplavcic — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| upgrade | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

File: pkg/apis/psmdb/v1/psmdb_defaults.go [hunk 4]
```diff
@@ -236,6 +206,48 @@ func (cr *PerconaServerMongoDB) CheckNSetDefaults(platform version.Platform, log
...
+				replset.Storage.WiredTiger.EngineConfig.CacheSizeRatio = defaultWiredTigerCacheSizeRatio
+			}
+			if replset.Storage.WiredTiger.IndexConfig == nil {
+				replset.Storage.WiredTiger.IndexConfig = &MongodSpecWiredTigerIndexConfig{
+					PrefixCompression: true,
+				}
+			}
+		}
+
+		if cr.Spec.Sharding.Enabled && replset.Storage.Engine == StorageEngineMMAPv1 {
```
Comment (delgod) [inline pkg/apis/psmdb/v1/psmdb_defaults.go:244]: we don't support it at all
```suggestion
		if replset.Storage.Engine == StorageEngineMMAPv1 {
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/apis/psmdb/v1/psmdb_defaults.go (hunk 4 above)
Comment (blez) [reply pkg/apis/psmdb/v1/psmdb_defaults.go]: https://github.com/percona/percona-server-mongodb-operator/pull/467/commits/7a642b9c66d5d6d42e70e5d8983f34bf461ad22d

**Review by delgod — APPROVED** [review]

**Review by tplavcic — APPROVED** [review]

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-467-16f175a0

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| operator-self-healing | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| service-per-pod | passed |
| scheduled-backup | passed |
| liveness | passed |
| demand-backup-sharded | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Review by nonemax — APPROVED** [review]

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-467-9f92f84a

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| self-healing | passed |
| limits | passed |
| operator-self-healing | passed |
| scaling | passed |
| one-pod | passed |
| security-context | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Pull request merged**